chainstate/src/detail/in_memory_block_tree.rs

```diff
@@ -0,0 +1,446 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use chainstate_storage::BlockchainStorageRead;
+use chainstate_types::{BlockIndex, InMemoryBlockTreeError};
+use common::{
+    chain::{block::timestamp::BlockTimestamp, Block},
+    primitives::{BlockHeight, Id},
+};
+use indextree::{Arena, Node, NodeId};
+use utils::{debug_panic_or_log, log_error};
+
+use crate::{detail::chainstateref::block_validity_matches, TransactionVerificationStrategy};
+
+use super::chainstateref::{BlockValidity, ChainstateRef};
+
+/// Zero or more `BlockIndex` trees sharing the same arena.
```

Comment:
Please explain why this is needed and what problem it solves

Reply:
To be clear, intuitively, multiple trees should go into a `BTreeMap<root,arena>`. Would be great to explain unintuitive things.